File: Excel/462 Challenge.R
library(tidyverse)
library(readxl)

input = read_excel("Excel/462 Fill in the Grid.xlsx", range = "A2:J11", col_names = F) %>%
  as.matrix()
test  = read_excel("Excel/462 Fill in the Grid.xlsx", range = "A14:J23", col_names = F) %>%
  as.matrix()

na_coords = which(is.na(input), arr.ind = T) 

get_surrounding_values = function(x, y, matrix){
  values = c()
  for (i in -1:1) {
    for (j in -1:1) {
      if (x + i > 0 & x + i <= nrow(matrix) & y + j > 0 & y + j <= ncol(matrix)) {
        values = c(values, matrix[x + i, y + j])
      }
    }
  }
  return(max(values, na.rm = T))
}


for (i in 1:nrow(na_coords)) {
  input[na_coords[i, 1], na_coords[i, 2]] = get_surrounding_values(na_coords[i, 1], na_coords[i, 2], input)
}

identical(input, test)
#> [1] TRUE

File: Excel/400-499/462/462 Challenge.py
import pandas as pd

input = pd.read_excel("462 Fill in the Grid.xlsx", sheet_name="Sheet1", header=None, skiprows=1, usecols="A:J", nrows = 10).values
test = pd.read_excel("462 Fill in the Grid.xlsx", sheet_name="Sheet1", header=None, skiprows=13, usecols="A:J", nrows = 10).values

na_coords = [(i, j) for i in range(len(input)) for j in range(len(input[0])) if pd.isna(input[i, j])]

def get_surrounding_values(x, y, matrix):
    values = []
    for i in range(-1, 2):
        for j in range(-1, 2):
            if 0 <= x+i < len(matrix) and 0 <= y+j < len(matrix[0]):
                values.append(matrix[x+i, y+j])
    return max(values, default=None)

for x, y in na_coords:
    input[x, y] = get_surrounding_values(x, y, input)

print((input == test).all()) # True